File: 287. Find the Duplicate Number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int slow = nums[0], fast = nums[nums[0]];
        while (slow != fast) {
            slow = nums[slow];
            fast = nums[nums[fast]];
        }
        fast = 0;
        while (slow != fast) {
            slow = nums[slow];
            fast = nums[fast];
        }

        return fast;
    }
};

// 重复数字就是环的入口， 将题目转换为求链表环的入口

File: go/src/leetcode/287. Find the Duplicate Number/287. Find the Duplicate Number.go
package main

import "fmt"

func findDuplicate(nums []int) int {
	var slow = nums[0]
	var fast = nums[slow]
	for slow != fast {
		slow = nums[slow]
		fast = nums[nums[fast]]
	}
	fast = 0
	for slow != fast {
		slow = nums[slow]
		fast = nums[fast]
	}
	return slow
}

func main() {
	fmt.Println(findDuplicate([]int{1,3,4,2,2}))
	fmt.Println(findDuplicate([]int{3,1,3,4,2}))
	fmt.Println(findDuplicate([]int{1,1}))
	fmt.Println(findDuplicate([]int{1,1,2}))
}


